include/graphblas/base/internalops.hpp

```diff
@@ -39,9 +39,127 @@ namespace grb {
 
 	namespace operators {
 
+
+
 		/** Core implementations of the standard operators in #grb::operators. */
 		namespace internal {
 
+			/**
+			 * Standard negation operator.
+			 *
+			 * Assumes native availability of ! on the given data types or assumes that
+			 * the relevant operators are properly overloaded.
+			 *
+			 * @tparam Op The Operator class to negate.
+			 * 			Requires the following typedefs:
+			 * 			- \b D1: The left-hand input domain.
+			 * 			- \b D2: The right-hand input domain.
+			 * 			- \b D3: The output domain, must be convertible to bool.
+			 * 			- \b operator_type: The internal::operator type to negate.
+			 */
+			template<
+				class Op,
+				enum Backend implementation = config::default_backend
+			>
+			class logical_not {
+				public:
+
+					/** Alias to the left-hand input data type. */
+					typedef typename Op::D1 left_type;
+
+					/** Alias to the right-hand input data type. */
+					typedef typename Op::D2 right_type;
+
+					/** Alias to the output data type. */
+					typedef typename Op::D3 result_type;
+
+					/** Whether this operator has an inplace foldl. */
+					static constexpr bool has_foldl = Op::operator_type::has_foldl;
+
+					/** Whether this operator has an inplace foldr. */
+					static constexpr bool has_foldr = Op::operator_type::has_foldr;
+
+					/**
+					 * Whether this operator is \em mathematically associative; that is,
+					 * associative when assuming equivalent data types for \a IN1, \a IN2,
+					 * and \a OUT, as well as assuming exact arithmetic, no overflows, etc.
+					 */
+					static constexpr bool is_associative = Op::operator_type::is_associative;
+
+					/**
+					 * Whether this operator is \em mathematically commutative; that is,
+					 * commutative when assuming equivalent data types for \a IN1, \a IN2,
+					 * and \a OUT, as well as assuming exact arithmetic, no overflows, etc.
+					 */
+					static constexpr bool is_commutative = Op::operator_type::is_commutative;
```

Comment:
By contrast, if the underlying operator is commutative then for sure the negated operator is commutative also, simply because no matter the order of operands, the negation is applied to the same value. Given the non-triviality of the associative case, it would be good to note why commutativity is "inherited" in the doxygen here.